File: leetcode/22/c++/solution_2.cpp
/*
    22. Generate Parentheses
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    void get_solution(int n, size_t open_parentheses, size_t close_parentheses,
                      std::string &solution, std::vector<std::string> &output) {
        if (open_parentheses == n && close_parentheses == n) {
            output.push_back(solution);
            return;
        }

        if (open_parentheses < n) {
            solution.push_back('(');
            get_solution(n, open_parentheses + 1, close_parentheses, solution,
                         output);
            solution.pop_back();
        }
        if (open_parentheses > close_parentheses) {
            solution.push_back(')');
            get_solution(n, open_parentheses, close_parentheses + 1, solution,
                         output);
            solution.pop_back();
        }
    }
    vector<string> generateParenthesis(int n) {
        std::vector<std::string> output;
        std::string solution;
        solution.reserve(n * 2);
        get_solution(n, 0, 0, solution, output);
        return output;
    }
};

File: leetcode/1930/rust/solution_2.rs
/*
    1930. Unique Length-3 Palindromic Subsequences
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn count_palindromic_subsequence(s : String) -> i32 {
        let mut forward_intervals : [(usize, usize); 26] = [(s.len(), 0); 26];
        let mut backward_intervals : [usize; 26] = [0; 26];
        let s = s.as_bytes();
        for
            i in 0..s.len() {
                let char_b_i : usize = (s[i] - b 'a') as usize;
                forward_intervals[char_b_i] .0 =
                    forward_intervals[char_b_i] .0.min(i);
                forward_intervals[char_b_i] .1 = i;
            }
        let mut output : usize = 0;

        for
            i in(0..s.len()).rev() {
                let char_b_i : usize = (s[i] - b 'a') as usize;
                if (forward_intervals[char_b_i] .0 == i &&
                    i < forward_intervals[char_b_i] .1) {
                    let(first, last) = forward_intervals[char_b_i];
                    output +=
                        backward_intervals.iter()
                            .filter(| &&last_backward | first < last_backward &&
                                    last > last_backward)
                            .count();
                }
                backward_intervals[char_b_i] = i;
            }
        output as i32
    }
}